### ISSUE TITLE
NEW-ENH: Stitching grid-tiled overlaping datasets

### ISSUE THREAD

**Issue by sem-geologist** [hyperspy/hyperspy#1410]
Hyperspy have nicely implemented `stacking` mechanism,
but what about enlarging dataset in the same dimensions (x,y)?
As the lazy loading is getting near the finish, I started to think about implementing stitching.
This feature probably would be most useful not for every microscope configurations but only for SEM's having possibly to gather hypermaps in programable grid.
I know some Bruker Esprit system do that (1st Question (What other EDS systems allows such grid on SEM, and is the stage positions implemented already in hyperspy (at least in original_metadata)?)
I am thinking about some limited implementation of it (well, stacking is also limiting it's input):
* all tiles should be done at same magnification, resolution
* tiles gathered with well defined grid
* either hypermap tile should provide stage metainfo, or filename should have well defined indices of tile_grid (or some metada parameter??)

Ideas of implementation:
1. Tweaking overlap:
1.1. If format reader returns few signals - allow to chose either simple 2D image (i.e. BSE) or generated sum EDS image.
1.1*(If it is ZEISS, get rid at least half of overlap from left side of the tile)
1.2. account the tile gird, build the neighborhood image dict (left, right, top, bottom) for every tile
1.3. calculate the initial overlapping regions of images from stage and resolution OR
interactively allow to choose entry overlap (i.e. GUI with 3x3 partly transparent images + x, y sliders for stage (if not present), plain overlap %, or resolution (if not present))
1.4. Find the shift correction by phase-correlation (scipy? opencv have extremely effective/fast implementation) and build the correction matrix/dict for the grid.
1.5 normalize the correction matrix/dict, offload common parts of vectors in correction matrix either to beam resolution or stage step/coordinates.
1.6 correct image absolute position for every tile and calculate final dimensions for the stitched array.

Those steps looks simple and some successfully experimented with, however there is other issue which makes it not clear how to do it:
What kind of final array should that be?
is dask enought?
or should it be chunked hdf5? which one h5py, pytables?
How effective Hyperspy hdf5 file format is for BIG DATA?

Personally I tried pytables with its chunking and extremely effective blosc compression, can h5py somehow compare?

**Comment by francisco-dlp** [comment]
I think that this would be a very nice addition, not only for SEM but also for STEM and probably also for fields beyond microscopy.

Some thoughts:

``stack`` lets you stack over a given axis and, therefore, it supports "naive" stitching already. You are proposing a more and powerful approach. I think that, instead of trying to implement everything at once, it'll be good to break it in more approachable tasks. For example, simply honoring the position metadata (or the axes information, which should work e.g. for FEI datasets) when stacking would represent a big improvement. It could be smart enough to allow for missing patches. 

I imagine that most of the time people will be stitching files in manufacturer's format, not hdf5 files but, in any case, if h5py is a bottleneck for any reason it wouldn't be a problem to use pytables instead.

For stitching images there is an [skimage tutorial](https://github.com/scikit-image/skimage-tutorials/blob/master/lectures/adv3_panorama-stitching.ipynb). I have the feeling that this particular task is beyond the scope of HyperSpy, what do others think?

**Comment by sem-geologist** [comment]
Exactly: stack goes along one axis, but stitching is for tiled data-sets.

skimage tutorial is completely out of scope there, stitching through particle finding and comparison probably works very well on everyday cameras (phone, photo camera, ) and so such approach for panorama building probably is ok. Stitching the tiled microscopy data-sets does not need all that scruff, instead most efficiently is to use fft phase-correlation, as the images taken in grid, supposedly should not have different rotations or magnitudes.

Actually hyperspy I think uses this fft phase-correlation method in correcting shifts in stacks already, I could generalize existing functions to be usable also to stitching.

**Comment by francisco-dlp** [comment]
Sorry for not explaining myself. The point is that the base "tiling" feature is already in ``stack``, but it requires some extra cleverness to make it pleasant to use. To tile images with the current ``stack`` you can stack first the signals for each row  and afterwards stack the "stacked rows" e.g. ``stack([stack(row_signals, axis="x") for row_signals in rslist], axis="y")``. It'll be very cool if it could do it more simply using the metadata  e.g. ``stack(signal_list, tile="auto")``.

**Comment by sem-geologist** [comment]
but does it properly align the images?

**Comment by francisco-dlp** [comment]
No, it doesn't. The point that  I was trying to make is that only adding this extra feature to ``stack`` would be really useful once #1219 gets merged. Aligning the images could be a separate, more specific task.

Imagine that you are using the feature to load some tens of big spectrum images or scanning electron diffraction data. It would be wasteful to align these datasets on stacking. A more efficient way would be to stack them naively, analyse them to obtain e.g. chemical maps and, afterwards, if needed due to drift etc, stitch the maps together using a less naive approach. Does it make sense?

**Comment by sem-geologist** [comment]
I should get working prototype... maybe then will it make sense...

**Comment by to266** [comment]
@sem-geologist 

Here is a working example of what Francisco meant:
```python
In [1]: import hyperspy.api as hs
   ...: from sklearn import datasets
   ...: 

In [2]: digits = datasets.load_digits()

In [3]: ims = hs.signals.Signal2D(digits.images[:100,...])

In [4]: ims
Out[4]: <Signal2D, title: , dimensions: (100|8, 8)>

In [5]: iters = [iter(ims.split())] * 10

In [6]: groupsof10 = zip(*iters)

In [7]: res = hs.stack([hs.stack(a, axis=0) for a in groupsof10], axis=1)

In [8]: res
Out[8]: <Signal2D, title: , dimensions: (|80, 80)>
```
I assume tiled datasets have some overlap between neighboring tiles. I agree with @francisco-dlp proposal to keep the alignment and stacking / tiling separate as much as possible, since these are actually quite different tasks. 

So, assuming some PR adds a function that takes a signal (image?) list and returns iterators (or a list of lists..) with properly cropped and ordered signals, the current `stack` is in principle able to do the rest.

**Comment by SANMoya** [comment]
Hi everyone, 
I am new to Hyperspy/Python coding and would like to know if there is a module for aligning focal series TEM images? 
this is the process I follow:
a= hs.load("Aligned_Stack.dm3")
SubPixAl= a.align2D()
When I say SubPixAl.plot()
It returns the following error: 
'numpy.ndarray' object has no attribute 'plot'

What am I doing wrong and how do I fix it?

Your response would be highly appreciated.
Thank you

**Comment by francisco-dlp** [comment]
The issue is that ``align2D`` does not returns the result (just the shifts). Instead it aligns ``a`` "in place", i.e. after calling ``align2D`` ``a`` should be aligned.

``align2D`` is a general registration function with some advanced functionality, such as Bernhard Schaffer's statistical approach. Does it work for your purposes?

By the way, and easier and usually faster way of asking questions to the hyperspy community is the [gitter chat](https://gitter.im/hyperspy/hyperspy)
